eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php

```diff
@@ -0,0 +1,161 @@
+<?php
+
+namespace eZ\Publish\Core\Persistence\Cache;
+
+use eZ\Publish\SPI\Persistence\Content\UrlWildcard\Handler as UrlWildcardHandlerInterface;
+
+class UrlWildcardHandler extends AbstractHandler implements UrlWildcardHandlerInterface
+{
+    /**
+     * @see \eZ\Publish\SPI\Persistence\Content\UrlWildcard\Handler::create
+     */
+    public function create($sourceUrl, $destinationUrl, $forward = false)
+    {
+        $this->logger->logCall(
+            __METHOD__,
+            [
+                'sourceUrl' => $sourceUrl,
+                'destinationUrl' => $destinationUrl,
+                'forward' => $forward,
+            ]
+        );
+
+        $urlWildcard = $this->persistenceHandler->urlWildcardHandler()->create($sourceUrl, $destinationUrl, $forward);
+
+        // need to clear lists of UrlWildcards cached due to loadAll()
+        $this->cache->invalidateTags(['ez-urlWildcard-all']);
+
+        return $urlWildcard;
+    }
+
+    /**
+     * @see \eZ\Publish\SPI\Persistence\Content\UrlWildcard\Handler::remove
+     */
+    public function remove($id)
+    {
+        $this->logger->logCall(
+            __METHOD__,
+            [
+                'id' => $id,
+            ]
+        );
+
+        $urlWildcard = $this->load($id);
+
+        $this->persistenceHandler->urlWildcardHandler()->remove($urlWildcard->id);
+
+        // need to clear lists of UrlWildcards cached due to loadAll()
+        $this->cache->invalidateTags(['ez-urlWildcard-all']);
+    }
+
+    /**
+     * @see \eZ\Publish\SPI\Persistence\Content\UrlWildcard\Handler::load
+     */
+    public function load($id)
+    {
+        $this->logger->logCall(
+            __METHOD__,
+            [
+                'id' => $id,
+            ]
+        );
+
+        $cacheItem = $this->cache->getItem('ez-urlWildcard-id-' . $id);
+
+        if ($cacheItem->isHit()) {
+            return $cacheItem->get();
+        }
+
+        $urlWildcard = $this->persistenceHandler->urlWildcardHandler()->load($id);
+
+        $cacheItem->set($urlWildcard);
+        $cacheItem->tag($this->getCacheTags([$urlWildcard]));
+        $this->cache->save($cacheItem);
+
+        $cacheItemBySource = $this->cache->getItem('ez-urlWildcard-source-' . $this->escapeForCacheKey($urlWildcard->sourceUrl));
+        $cacheItemBySource->set($urlWildcard);
+        $cacheItem->tag($this->getCacheTags([$urlWildcard]));
+        $this->cache->save($cacheItemBySource);
+
+        return $urlWildcard;
+    }
+
+    /**
+     * @see \eZ\Publish\SPI\Persistence\Content\UrlWildcard\Handler::loadAll
+     */
+    public function loadAll($offset = 0, $limit = -1)
+    {
+        $this->logger->logCall(
+            __METHOD__,
+            [
+                'offset' => $offset,
+                'limit' => $limit,
+            ]
+        );
+
+        $cacheItem = $this->cache->getItem('ez-urlWildcard-all-' . $offset . '-' . $limit);
```

Comment:
I'll rather drop caching load all, it can be a very large result. _So assuming we fix the design here there won't be a need to pre-emptive try to optimize a load all_

Reply:
ref legacy comment above, maybe the answer is not to load all, but rather expose a ending to load all in a compact format that can be effectively used by regex matching.

Unless someone sees how we can do the match in sql _(picking first match by order)_, and cache the results.